### PR TITLE
Fix memory leak in AF setting/clearing functions

### PR DESCRIPTION
**Pull request by jimrandomh**
Fix a memory leak which periodically kills causes LW server instances to die, causing 504s. The problem is that `updateParentsSetAFtrue` can recurse infinitely if searching for a null `parentCommentId` returns a result, leaking at a rate of one stack-frame per mongodb roundtrip until the process dies.

Fix rewrites `updateParentsSetAFtrue` (and `updateChildrenSetAFfalse`, which was right next to it and looked a little sketchy) to use preexisting, safer utility functions for getting the comment ancestor-chain and descendents. (Then moves those utility functions into a new file to break an import-cycle, and armors `getCommentAncestorIds` against the possibility of parent-comment-reference cycles in the database.)

This bug was introduced in https://github.com/ForumMagnum/ForumMagnum/commit/e7d2dbf184249e30d2d14974322cc3faca3ec617 and has likely existed ever since, but may have been masked for most of its lifetime by fortunate coincidences about which comment gets returned when you make an unconstrained find-a-comment query.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1202944851758246) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/utils/commentTreeUtils.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  let unvisited: string[] = [rootComment._id];
+  
+  while(unvisited.length > 0) {
+    const childComments = await Comments.find({parentCommentId: {$in: unvisited}}, projection).fetch();
+    for (let commentId of unvisited)
+      visited.add(commentId);
+    unvisited = [];
+    
+    for (let childComment of childComments) {
+      if (!visited.has(childComment._id)) {
```
Comment (Xodarap) [inline packages/lesswrong/server/utils/commentTreeUtils.ts:37]: I slightly think that throwing an error here as well would be good; it makes debugging easier (at the expense of the callback not working, of course)

**Review by Xodarap — APPROVED** [review]
Maybe it's worth updating the mutation checks to ensure that we can't introduce a cycle in the future (I assume this will cause problems in numerous places in the code base, not just here).

I also made one other small suggestion, but lgtm :shipit:

File: packages/lesswrong/server/utils/commentTreeUtils.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import { Comments } from '../../lib/collections/comments/collection';
+import * as _ from 'underscore';
```
Comment (b0b3rt) [inline packages/lesswrong/server/utils/commentTreeUtils.ts:2]: don't think the `underscore` import is used here

**Review by b0b3rt — APPROVED** [review]
Apart from an unused import, looks good to me

**Pull request merged**